Add pressure metric for bpf maps

jcaamano opened the pull request:
Adds a metric to tell the fill percentage from 0 to 1 of selected maps
filled from userspace. Selected maps are those non LRU filled from
userspace.

This metric is a Prometheus const label gauge that is registered and
unregistered dynamically for each map when over/under a threshold
respectively. By default the threshold is 0.0 for all tracked maps.

As other metrics, this one can be enabled or disabled globally via
command line and is disabled by default.

This metric takes advantage of the existing cache mechanism for maps
to count the current number of entries on each map. For those maps that
already have this cache enabled, there is no change in this regard. For
those that don't, this caching mechanism is partially enabled to track a
key-only cache, which is the case for ` cilium_lb4_reverse_sk`,
`cilium_lb6_reverse_sk`, `cilium_lb4_maglev` and `cilium_lb6_maglev`
maps.

The policy map is an exceptional case. For this map threshold is set to
0.1 to avoid flooding Prometheus with reports. The metric is tracked
from the endpoint instead of the map in this case as the endpoint is
already caching this data while the map does not.

**Release note**
```release-note
Add new `cilium_bpf_map_pressure` metric measuring the fill-up ratio of selected BPF maps.
```

jcaamano committed:
Add pressure metric for bpf maps

Adds a metric to tell the fill percentage from 0 to 1 of selected maps
filled from userspace. Selected maps are those non LRU filled from
userspace.

This metric is a Prometheus const label gauge that is registered and
unregistered dynamically for each map when over/under a threshold
respectively. By default the threshold is 0.0 for all tracked maps.

As other metrics, this one can be enabled or disabled globally via
command line and is disabled by default.

This metric takes advantage of the existing cache mechanism for maps
to count the current number of entries on each map. For those maps that
already have this cache enabled, there is no change in this regard. For
those that don't, this caching mechanism is partially enabled track a
key-only cache, which is the case for `cilium_lb4_reverse_sk`,
`cilium_lb6_reverse_sk`, `cilium_lb4_maglev` and `cilium_lb6_maglev`
maps.

The policy map is an exceptional case. For this map threshold is set to
0.1 to avoid flooding prometheus with reports. The metric is tracked
from the endpoint instead of the map in this case as the endpoint is
already caching this data while the map does not.

Signed-off-by: Jaime Caamaño Ruiz <jcaamano@suse.com>

aanm commented on pkg/bpf/map_linux.go:
```diff
@@ -214,6 +225,10 @@ func (m *Map) commonName() string {
 }
 
+func (m *Map) nonPrefixedName() string {
+	return m.name[7:]
```
and why `7`?
```
if len(m.name) >= 7 {
return m.name[7:]
}
return m.name
```

jcaamano replied:
Removing `cilium_` prefix but unlike it is done with the common name we want to keep things like the endpoint id in the name.

joestringer replied:
I would suggest using something like `strings.TrimPrefix(...)` to prevent unintended behaviour here.

vadorovsky commented on pkg/bpf/map_linux.go:
```diff
@@ -117,7 +117,10 @@ type Map struct {
 	// DumpParser is a function for parsing keys and values from BPF maps
 	DumpParser DumpParser
 
-	cache map[string]*cacheEntry
```
I don't think you have to change the type of `cache`. For empty entries we can just use `nil` (with an explicit assignment, like you already do, so the key would exists and we could look up for it). And we could avoid type conversions.

jcaamano replied:
Right, I was using a different empty struct type before but forgot to change this back.

vadorovsky committed:
grafana: Add BPF map pressure chart

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
helm: Add prometheus.metrics value

The new prometheus.metrics value allows to disable and enable Prometheus
metrics (+metric_foo to enable metric_foo, -metric_bar to enable
metric_bar) and is used in the ConfigMap.

Example of usage:

  helm install cilium cilium/cilium \
    --namespace=kube-system \
    --set prometheus.enabled=true \
    --set prometheus.metrics="{-cilium_bpf_map_ops_total,+cilium_bpf_map_pressure}" \
    --set operator.prometheus.enabled=true

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky commented on pkg/metrics/metrics.go:
```diff
@@ -553,6 +555,7 @@ func DefaultMetrics() map[string]struct{} {
 		Namespace + "_" + SubsystemKVStore + "_quorum_errors_total":                  {},
 		Namespace + "_fqdn_gc_deletions_total":                                       {},
 		Namespace + "_" + SubsystemBPF + "_map_ops_total":                            {},
+		Namespace + "_" + SubsystemBPF + "_map_pressure":                             {},
```
To make it optional, maybe let's remove it from here and apply the following Helm change I made locally:

https://github.com/mrostecki/cilium/commit/bcd661b90683ace4f637e4bf2b51b962f7d2a895

Feel free to just pull/am/cherry-pick from my branch, if my changes look good to you (or modify them if they don't), there is also a Grafana change:

https://github.com/mrostecki/cilium/commits/map-pressure-metric

jcaamano replied:
I removed the metric from the defaults and added your commits to this PR. The helm one might have merit for it's own PR though.

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -863,6 +863,10 @@ prometheus:
     # -- Specify the Kubernetes namespace where Prometheus expects to find
     # service monitors configured.
     # namespace: ""
+  # Metrics that should be enabled or disabled from the default metric
```
Couple of things here:

nit, formatting like this (like the other options above) allows the automated tooling we use on these values to pick up the descriptions and properly format them:
```suggestion
  # -- Metrics that should be enabled or disabled from the default metric
```

Additionally, would it be reasonable to also document what the valid options are here? Or is that better documented somewhere else?

jcaamano replied:
Added a reference to the docs.

qmonnet commented on pkg/metrics/metrics_test.go:
```diff
@@ -0,0 +1,71 @@
+// Copyright 2019 Authors of Cilium
```
Nit: Copyright year should be 2020?

qmonnet commented on pkg/metrics/metrics_test.go:
```diff
@@ -0,0 +1,71 @@
+	c.Assert(err, IsNil)
+	c.Assert(metrics, HasLen, initMetricLen)
+	c.Assert(GetGaugeValue(gauge.gauge), Equals, underThreshold)
+
```
Nit: unnecessary blank line.

qmonnet reviewed (changes requested):
Changes to the documentation look good.

I only had a quick glance at the changes in the agent - Please find some minor nits below.

qmonnet reviewed (approved):
Thanks a lot!

kaworu reviewed (approved):
Helm related change LGTM

aanm commented on pkg/bpf/map_linux.go:
```diff
@@ -245,11 +257,51 @@ func (m *Map) scheduleErrorResolver() {
+// threshold.
+func (m *Map) WithPressureMetricThreshold(threshold float64) *Map {
+	// When pressure metric is enabled, we keep track of map keys in cache
+	if m.cache == nil {
+		m.cache = map[string]*cacheEntry{}
+	}
```
Maybe would be better to call `m.WithCache()` here? otherwise we will have `m.cache` with data but `m.withCache` will be false.

jcaamano replied:
`m.cache` and `m.WithCache()`/`m.withCache` are different things. `m.cache` is used also for the pressure metric key-only cache unless `m.withCache` flag is set in which case is a key-value cache.  Perhaps renaming `m.withCache` to `m.withValueCache` is a bit more clear.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -1071,6 +1073,19 @@ func (e *Endpoint) SkipStateClean() {
+
+	if e.policyMapPressureGauge == nil {
+		e.policyMapPressureGauge = metrics.NewBPFMapPressureGauge(e.policyMap.NonPrefixedName(), policymap.PressureMetricThreshold)
+	}
```
This is something that can potentially be in the endpoint constructor.

jcaamano replied:
Yes. I wasn't sure if there was a single endpoint constructor or many. Let me try to track them down, please check after.

jcaamano replied:
I moved this to be constructed with the policyMap which is deferred from the endpoint constructor as well but at least we are not checking for it on each metric value set.

joestringer commented on pkg/bpf/map_linux.go:
```diff
@@ -880,6 +938,8 @@ func (m *Map) deleteCacheEntry(key MapKey, err error) {
 	k := key.String()
 	if err == nil {
 		delete(m.cache, k)
+	} else if !m.withValueCache {
+		return
```
Why is this necessary? Line 934 already checks `if m.cache == nil { return }`.

jcaamano replied:
`m.cache` is used both as a key-only cache for maps with pressure metric enabled but caching disabled, or as a key/value cache for maps with caching enabled. Thus `m.cache != nil` no longer tracks whether map caching is enabled, `withValueCache` does.

joestringer commented on pkg/bpf/map_linux.go:
```diff
@@ -117,6 +118,9 @@ type Map struct {
 	// DumpParser is a function for parsing keys and values from BPF maps
 	DumpParser DumpParser
 
+	// withValueCache is true when map cache has been enabled
+	withValueCache bool
```
Is this different than `m.cache != nil`? Tracking an extra variable seems like provides an opportunity for the bool vs. the pointer to get out of sync.

If there is an actual difference, that would be useful context to put in the comment above this field.

jcaamano replied:
Added comment on `m.cache` member that hopefully clarifies.

joestringer commented on pkg/bpf/map_linux.go:
```diff
@@ -933,7 +994,7 @@ func (m *Map) DeleteAll() error {
 	nextKey := make([]byte, m.KeySize)
 
-	if m.cache != nil {
+	if m.withValueCache {
```
I couldn't figure out, does `DeleteAll()` update the metric?

jcaamano replied:
It didn't. Added.

joestringer commented on pkg/bpf/map_linux.go:
```diff
@@ -836,22 +888,28 @@ func (m *Map) Update(key MapKey, value MapValue) error {
 	m.lock.Lock()
 	defer m.lock.Unlock()
 
+	defer m.updatePressureMetric()
```
I'd guess we only need to update this metric if `err != nil`, any particular reason not to have it inside the same defer function below which already has checks for whether the cache is enabled and checks against the `err`?

jcaamano replied:
I removed the unnecessary `defer` and added `updatePressureMetric()` to the existing func. Still called when `err != nil` and caching enabled since even in that case an entry is added to the cache.
Also, I was calling `updatePressureMetric()` pretty liberally due to its idempotency but I tried to tweak better the places where it is called.

joestringer commented on pkg/bpf/map_linux.go:
```diff
@@ -245,11 +257,51 @@ func (m *Map) scheduleErrorResolver() {
+		return
+	}
+
+	if !option.Config.MetricsConfig.BPFMapPressure {
+		if !m.withValueCache {
+			m.cache = nil
+		}
+		m.pressureGauge = nil
+		return
+	}
```
[0] Why would `m.pressureGauge` be non-nil at this point if `!option.Config.MetricsConfig.BPFMapPressure`? I would have thought that this field should be initialized to `nil` during the very first attempt to configure the gauge rather than checked & possibly disabled any time there is a map update.

I don't think we have any code that might cause `option.Config.MetricsConfig.BPFMapPressure()` to change at runtime so it should be safe to move it there.

jcaamano replied:
On one side we cannot check `option.Config.MetricsConfig.BPFMapPressure` on initialization due to the static initialization of the maps when `option.Config` has not been loaded yet.

`m.pressureGauge` is initialized non-nil. While it could be initialized nil I chose to do it similar and inline with `m.cache` which cannot be initialized nil and which actually allows for less checks.

joestringer replied:
Ah, static initialization and argument parsing strikes again. Thanks for the explanation.

This kind of detail would actually be nice to leave as a comment above the option check to ensure that in future we don't just assume we can move it and rely on the results (unless some other sync mechanism is put in place to ensure the options are available).

jcaamano replied:
Added comment!

christarazi replied:
Is this PR (https://github.com/cilium/cilium/pull/14607) related? Is this what's meant by "static initialization"?

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1071,6 +1074,27 @@ func (e *Endpoint) SkipStateClean() {
+	e.policyMapPressureGauge = metrics.NewBPFMapPressureGauge(e.policyMap.NonPrefixedName(), policymap.PressureMetricThreshold)
+}
+
+func (e *Endpoint) updatePolicyMapPressureMetric() {
+	if e.policyMapPressureGauge == nil {
+		return
+	}
+
+	value := float64(len(e.realizedPolicy.PolicyMapState)) / float64(e.policyMap.MapInfo.MaxEntries)
+	e.policyMapPressureGauge.Set(value)
+}
```
One of the reasons I was drawn to look at [0] was that I noticed this function appears to be equivalent but different in implementation due to the code I pointed out at [0]. Ideally these should be consistent (Ideally using the same object definition, shared across both users; but there's only a few lines so that's not such a big deal).

Would you mind either consistently doing the config check in the init() or in the update() across both implementations?

jcaamano replied:
Here is different because we can check `option.Config.MetricsConfig.BPFMapPressure` on initialization which saves doing extra checks in the update.

joestringer reviewed (changes requested):
Thanks for working on this, and sorry about the delay in review :-)

Looks good overall, just a few minor nits here and there.

joestringer reviewed (approved):
Would be nice to add the explanation for why the check is there on every metric update into a comment but otherwise LGTM :+1:

vadorovsky reviewed (approved):
ACK for @jcaamano's changes

aanm reviewed (commented):
@jcaamano the CI is failing and the errors are related with eBPF maps. I don't think these are flakes.
